core/message_queue.go

```diff
@@ -0,0 +1,131 @@
+package core
+
+import (
+	"sync"
+
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/pkg/errors"
+)
+
+// MessageQueue stores an ordered list of messages (per actor) and enforces that their nonces form a contiguous sequence.
+// Each message is associated with a "stamp" (an opaque integer), and the queue supports expiring any list
+// of messages where the first message has a stamp below some threshold. The relative order of stamps in a queue is
+// not enforced.
+// A message queue is intended to record outbound messages that have been transmitted but not yet appeared in a block,
+// where the stamp could be block height.
+// MessageQueue is safe for concurrent access.
+type MessageQueue struct {
+	lk sync.RWMutex
+	// Message queues keyed by sending actor address, in nonce order
+	queues map[address.Address][]*queuedMessage
+}
+
+type queuedMessage struct {
+	msg   *types.SignedMessage
+	stamp uint64
+}
+
+// NewMessageQueue constructs a new, empty queue.
+func NewMessageQueue() *MessageQueue {
+	return &MessageQueue{
+		queues: make(map[address.Address][]*queuedMessage),
+	}
+}
+
+// Enqueue appends a new message for an address. If the queue already contains any messages for
+// from same address, the new message's nonce must be exactly one greater than the largest nonce
+// present.
+func (mq *MessageQueue) Enqueue(msg *types.SignedMessage, stamp uint64) error {
+	mq.lk.Lock()
+	defer mq.lk.Unlock()
+
+	q := mq.queues[msg.From]
+	if len(q) > 0 {
+		nextNonce := q[len(q)-1].msg.Nonce + 1
+		if msg.Nonce != nextNonce {
+			return errors.Errorf("Invalid nonce %d, expected %d", msg.Nonce, nextNonce)
+		}
+	}
+	mq.queues[msg.From] = append(q, &queuedMessage{msg, stamp})
+	return nil
+}
+
+// RemoveNext removes and returns a single message from the queue, if it bears the expected nonce value.
+// Returns the message removed, or nil if the queue is empty or the expected nonce was less than any
+// in the queue for that address (indicating the message had already been removed).
+// Returns an error if the expected nonce is greater than the smallest in the queue.
+// The caller may wish to check that the returned message is equal to that expected (not just in nonce value).
+func (mq *MessageQueue) RemoveNext(sender address.Address, expectedNonce uint64) (*types.SignedMessage, error) {
+	mq.lk.Lock()
+	defer mq.lk.Unlock()
+
+	q := mq.queues[sender]
+	if len(q) > 0 {
+		head := q[0]
+		if expectedNonce < uint64(head.msg.Nonce) {
+			return nil, nil
+		} else if expectedNonce == uint64(head.msg.Nonce) {
+			mq.queues[sender] = q[1:] // pop the head
+			return head.msg, nil
+		}
+		return nil, errors.Errorf("Next message for %s has nonce %d, expected %d", sender, head.msg.Nonce, expectedNonce)
+	}
+	return nil, nil
```

Comment:
I'd rather this function return special not found error or add an `ok` return value. Not a big deal, and all the options are a little ugly, but signalling by returning nils is not idiomatic in Go.